packages/url_launcher/url_launcher_platform_interface/lib/link.dart

```diff
@@ -0,0 +1,99 @@
+
+/// Signature for a function provided by the [Link] widget that instructs it to
+/// follow the link.
+typedef FollowLink = Future<void> Function();
+
+/// Signature for a builder function passed to the [Link] widget to construct
+/// the widget tree under it.
+typedef LinkWidgetBuilder = Widget Function(
+  BuildContext context,
+  FollowLink followLink,
+);
+
+/// Signature for a delegate function to build the [Link] widget.
+typedef LinkDelegate = Widget Function(LinkInfo linkWidget);
+
+final MethodCodec _codec = const JSONMethodCodec();
+
+/// Defines where a Link URL should be open.
+///
+/// This is a class instead of an enum to allow future customizability e.g.
+/// opening a link in a specific iframe.
+class LinkTarget {
+  /// Const private constructor with a [debugLabel] to allow the creation of
+  /// multiple distinct const instances.
+  const LinkTarget._({this.debugLabel});
+
+  /// Used to distinguish multiple const instances of [LinkTarget].
+  final String debugLabel;
+
+  /// Use the default target for each platform.
+  ///
+  /// On Android, the default is [blank]. On the web, the default is [self].
+  ///
+  /// iOS, on the other hand, defaults to [self] for web URLs, and [blank] for
+  /// non-web URLs.
+  static const defaultTarget = LinkTarget._(debugLabel: 'defaultTarget');
+
+  /// On the web, this opens the link in the same tab where the flutter app is
+  /// running.
+  ///
+  /// On Android and iOS, this opens the link in a webview within the app.
+  static const self = LinkTarget._(debugLabel: 'self');
+
+  /// On the web, this opens the link in a new tab or window (depending on the
+  /// browser and user configuration).
+  ///
+  /// On Android and iOS, this opens the link in the browser or the relevant
+  /// app.
+  static const blank = LinkTarget._(debugLabel: 'blank');
+}
+
+/// Used to override the delegate that builds the link.
+LinkDelegate linkDelegate;
+
+/// Encapsulates all the information necessary to build a Link widget.
+abstract class LinkInfo {
+  /// Called at build time to construct the widget tree under the link.
+  LinkWidgetBuilder get builder;
+
+  /// The destination that this link leads to.
+  Uri get uri;
+
+  /// The target indicating where to open the link.
+  LinkTarget get target;
+
+  /// Whether the link is disabled or not.
+  bool get isDisabled;
+}
+
+/// Pushes the [routeName] into Flutter's navigation system via a platform
+/// message.
+Future<ByteData> pushRouteNameToFramework(
+  String routeName, {
+  @required bool isUsingRouter,
+}) {
+  final MethodCall call = isUsingRouter
+      ? MethodCall('pushRouteInformation', <dynamic, dynamic>{
+          'location': routeName,
+          'state': null,
+        })
+      : MethodCall('pushRoute', routeName);
+  final Completer<ByteData> completer = Completer<ByteData>();
+  window.onPlatformMessage(
+    'flutter/navigation',
+    _codec.encodeMethodCall(call),
+    completer.complete,
+  );
```

Comment:
Is there an imperative Router API that can be called, instead of platform messages?

Reply:
That's a great idea actually.

@chunhtai if I get the Router instance from `Router.of(context)`, is there a way to push the route information name without sending a platform message?

Maybe I can do the same thing with `Navigator.of(context).pushNamed(...)` when there's no Router?

Reply:
There is no imperative RouterAPI for route update, but you should use SystemNavigator.routeInformationUpdated or SystemNavigator.routeUpdated

Reply:
is the purpose of this function to just update the browser URL bar? calling this method alone will not make navigator/router update their content.

Reply:
The purpose of this function is to tell the framework to push the given route name. This is invoked when a `Link` widget is pressed. Consequently, the browser URL bar will be updated when the framework sends the `routeUpdated`/`routeInformationUpdated` message to the engine.

Reply:
Would it work if I do:
```dart
final Router router = Router.of(context);
if (router != null) {
  router.routeInformationProvider.value = {'location': routeName, 'state': null};
} else {
  Navigator.pushNamed(context, routeName);
}
```
?

Reply:
It wouldn't work for either case, for router the routeInformationProvider is a value listenable, it does not have a setter.
For navigator, it wouldn't work if it has nested navigator, you can probably get the root navigator, but it is not systematic correct.

The best way to do this is somehow trigger didpushRoute/didPushRouteInformation on widgetbindingobserver.
This can be done either by exposing some WidgetBinding API, or send a message through the message channel in the web engine. Besides that you still need to send SystemNavigator.routeInformationUpdated or SystemNavigator.routeUpdated.

Reply:
Ok I got both to work now. For Navigator, I'm just sending a "pushRoute" message to the framework (which goes through binding's `didPushRoute`). For Router, like you mentioned, I had to send a "pushRouteInformation" message to the framework (which goes through binding's `didPushRouteInformation`) AND call `SystemNavigator.routeInformationUpdated` to update the URL in the browser.